hibernate-core/src/main/java/org/hibernate/loader/ast/internal/LoaderSelectBuilder.java

```diff
@@ -540,6 +547,12 @@ else if ( fetchDepth > maximumFetchDepth ) {
 						null,
 						creationState
 				);
+				if ( fetch.getTiming() == FetchTiming.IMMEDIATE && fetchable instanceof PluralAttributeMapping ) {
+					if ( ( (PluralAttributeMapping) fetchable ).getMappedTypeDescriptor()
+							.getCollectionSemantics() instanceof BagSemantics ) {
+						bagRoles.add( fetchable.getNavigableRole().getNavigableName() );
+					}
+				}
 				fetches.add( fetch );
 
 				if ( fetchable instanceof PluralAttributeMapping && fetchTiming == FetchTiming.IMMEDIATE && joined ) {
```

Comment:
It seems the above two if blocks could be combined as following to boost some performance:
```
if ( fetch.getTiming() == FetchTiming.IMMEDIATE && fetchable instanceof PluralAttributeMapping ) {
    if ( ( (PluralAttributeMapping) fetchable ).getMappedTypeDescriptor()
							.getCollectionSemantics() instanceof BagSemantics ) {
	bagRoles.add( fetchable.getNavigableRole().getNavigableName() );
    }
    if ( joined ) {
        ... ...
    }
}
```